{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "updated drawing \n",
    "(change art style)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from typing import List,Tuple,Callable\n",
    "\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from math import log2\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import cm\n",
    "from matplotlib import gridspec\n",
    "import matplotlib.colors as mcolors\n",
    "from matplotlib.colors import LinearSegmentedColormap,Normalize,FuncNorm\n",
    "import matplotlib.patches as patches\n",
    "from matplotlib.axes import Axes\n",
    "from foreground import get_foreground \n",
    "from dna_features_viewer import GraphicFeature, GraphicRecord"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def parse_diamond(matchfile:str='diamond-matches.tsv')->pd.DataFrame:\n",
    "    '''\n",
    "    matchfile:\n",
    "        output of diamond blastp\n",
    "        with '--header', 'verbose', \n",
    "        and 'tsv' suffix\n",
    "    '''\n",
    "    head_lines=open(matchfile,'r').readlines()[2].strip().split(': ')[1]\n",
    "    diamond_aligns=pd.read_csv(matchfile,skiprows=3,names=head_lines.split(', '),delim_whitespace=True)\n",
    "    return diamond_aligns\n",
    "\n",
    "def get_identities(fasta_path:str):\n",
    "    '''\n",
    "    from diamond tsv\n",
    "    to df of ['domain','virus','identity']\n",
    "    '''\n",
    "    o_dict={}\n",
    "    _={}\n",
    "    _['domain'],_['virus'],_['identity']=[],[],[]\n",
    "    for i in Path(fasta_path).iterdir():\n",
    "        if i.suffix=='.tsv':\n",
    "            stem=i.stem.replace('-match','')\n",
    "            diamond=parse_diamond(i.absolute())\n",
    "            o_dict[stem]={i:j for i,j in zip(diamond['Subject ID'],diamond['Percentage of identical matches'])}\n",
    "            for idx,s in diamond.iterrows():\n",
    "                _['domain'].append(stem)\n",
    "                _['virus'].append(s['Subject ID'])\n",
    "                _['identity'].append(s['Percentage of identical matches'])\n",
    "    identities=pd.DataFrame(_)\n",
    "    # identities.columns\n",
    "    return o_dict,identities\n",
    "\n",
    "def cal_lr(x,y,refx=np.linspace(1,100,100)):\n",
    "    lr=LinearRegression()\n",
    "    x_=np.array(x).reshape(-1,1)\n",
    "    y_=np.array(y).reshape(-1,1)\n",
    "    reg=lr.fit(X=x_,y=y_)\n",
    "    r2=reg.score(x_,y_)\n",
    "    coef=reg.coef_[0][0]\n",
    "    intercept=reg.intercept_[0]\n",
    "    \n",
    "    predy=reg.predict(refx.reshape(-1,1))\n",
    "    sel=(predy>0) & (predy<100)\n",
    "    \n",
    "    reg_x=refx[sel.reshape(-1)].reshape(-1)\n",
    "    reg_y=predy[sel].reshape(-1)\n",
    "    \n",
    "    return (\n",
    "        coef,intercept,r2,\n",
    "        reg_x,reg_y,\n",
    "    )\n",
    "    \n",
    "def get_lr_cohort(identities_dict:dict,\n",
    "                  ref_genome:str,\n",
    "                  ref_accession:str):\n",
    "    '''use `identity_dict` from `get_identities`'''\n",
    "    o_dict={}\n",
    "    for accession,d in identities_dict.items():\n",
    "            # if accession!=ref_accession:\n",
    "            x,y=[],[]\n",
    "            for k in d.keys():\n",
    "                if k !=ref_genome:\n",
    "                    r=identities_dict[ref_accession].get(k,None)\n",
    "                    if r is not None:\n",
    "                        x.append(r)\n",
    "                        y.append(d[k])\n",
    "            if len(x)>0:\n",
    "                o_dict[accession]=(x,y)\n",
    "    return o_dict\n",
    "\n",
    "\n",
    "\n",
    "def get_ref_palette(accession:str,\n",
    "                    lr_metrics:dict,\n",
    "                    corr_color_map:LinearSegmentedColormap,\n",
    "                    fetch=lambda x: x[0],\n",
    "                    transform=lambda x: 1-log2(x)):\n",
    "    '''\n",
    "    `lr_metrics` comes from `{k:cal_lr(v[0],v[1]) for k,v in lr_cohort.items()}`,\n",
    "        where `lr_cohort` comes from get_lr_cohort\n",
    "    `fetch`: fetch metric values from lr_metrics' dict\n",
    "    `trasnform`: metric values mapping (usually to 0~1)\n",
    "    '''\n",
    "    if accession in lr_metrics:\n",
    "        # ref_c=corr_color_map(lr_metrics[accession][2]/100-1e-10)\n",
    "        m_val=transform(fetch(lr_metrics[accession]))\n",
    "        ref_c=corr_color_map(m_val)\n",
    "        return LinearSegmentedColormap.from_list(accession, [(1, 1, 1, 1), ref_c])\n",
    "    else:\n",
    "        return LinearSegmentedColormap.from_list(accession, [(1, 1, 1), (1, 0, 0)])\n",
    "    \n",
    "def get_missing_palette(domains:pd.DataFrame,ref_genome:str,\n",
    "        missing_color_map:LinearSegmentedColormap)->Tuple[dict[str,Tuple[int]],int]:\n",
    "    o_dict={}\n",
    "    # total_genome=len(domains['genome_name'].unique())\n",
    "    cano_entry=domains[domains['domain_accession']==ref_genome]['genome_name'].unique()\n",
    "    uncano_entry=[i for i in domains['domain_accession'].unique() \n",
    "                    if i not in cano_entry]\n",
    "    max_missing=max([len(domains[domains['domain_accession']==i]) for i in uncano_entry])\n",
    "    for i in uncano_entry:\n",
    "        l=len(domains[domains['domain_accession']==i])\n",
    "        o_dict[i]=missing_color_map(l/max_missing)\n",
    "    return o_dict,max_missing\n",
    "\n",
    "\n",
    "\n",
    "def get_corr_color_map(lr_metrics:dict,fetch=lambda x: x[0],\n",
    "                       colors = [mcolors.to_rgba(i) for i in ['teal','gold','orangered']]):\n",
    "    _=[fetch(i) for i in lr_metrics.values()]\n",
    "    metric_min,metric_max=min(_),max(_)\n",
    "    # print(metric_min,metric_max)\n",
    "    position_ref=(1-metric_min)/(metric_max-metric_min)\n",
    "    if position_ref<=0:\n",
    "        positions = [0, 1]\n",
    "        situ='conserved'\n",
    "        corr_color_map = mcolors.LinearSegmentedColormap.from_list(\n",
    "            f'{situ}_color_map', [(i,j) for i,j in zip(positions,colors[:2])], N=512)\n",
    "        \n",
    "    elif 0<position_ref<1:\n",
    "        positions = [0, position_ref, 1]\n",
    "        situ='biased'\n",
    "        corr_color_map = mcolors.LinearSegmentedColormap.from_list(\n",
    "            f'{situ}_color_map', [(i,j) for i,j in zip(positions,colors)], N=512)\n",
    "    else :\n",
    "        positions = [0, 1]\n",
    "        situ='diverged'\n",
    "        corr_color_map = mcolors.LinearSegmentedColormap.from_list(\n",
    "            f'{situ}_color_map', [(i,j) for i,j in zip(positions,colors[1:])], N=512)\n",
    "    return corr_color_map,situ,position_ref,metric_min,metric_max\n",
    "\n",
    "def fix_corr_color_map(lr_metrics:dict,fetch=lambda x: x[0],\n",
    "                       colors = [mcolors.to_rgba(i) for i in ['teal','gold','orangered']]):\n",
    "    situ='fix'\n",
    "    position_ref=1/3\n",
    "    metric_min,metric_max=0.5,2\n",
    "    corr_color_map = mcolors.LinearSegmentedColormap.from_list(\n",
    "            f'{situ}_color_map', [(i,j) for i,j in zip([0, position_ref, 1],colors)], N=512)\n",
    "    return corr_color_map,situ,position_ref,metric_min,metric_max\n",
    "\n",
    "def hide_axes(ax:Axes,genome_start_pos:int,genome_end_pos:int,sequence_length=42000):\n",
    "    '''\n",
    "    hack: draw a white box over them.\n",
    "    '''\n",
    "    ax.add_patch(\n",
    "     patches.Rectangle(\n",
    "        (-100, -0.25),\n",
    "        genome_start_pos,\n",
    "        0.5,\n",
    "        fill=True,      # remove background\n",
    "        facecolor='white'\n",
    "     ) ) \n",
    "    ax.add_patch(\n",
    "     patches.Rectangle(\n",
    "        (genome_end_pos, -0.25),\n",
    "        sequence_length+100,\n",
    "        0.5,\n",
    "        fill=True,      # remove background\n",
    "        facecolor='white'\n",
    "     ) ) \n",
    "\n",
    "def sort_genome(genome_names:List[str],sort_rdrps:pd.DataFrame):\n",
    "    o=[]\n",
    "    for i in genome_names:\n",
    "        if i in sort_rdrps.index:\n",
    "            o.append((i,sort_rdrps.loc[i]['Percentage of identical matches']))\n",
    "        else:\n",
    "            o.append((i,0))\n",
    "    o.sort(key=lambda x:x[1],reverse=True)\n",
    "    return o\n",
    "# def model_function(x:np.ndarray, a:float)->np.ndarray:\n",
    "#     return 2 ** (x ** a)-1\n",
    "\n",
    "# def cal_cur(x,y,model_function:Callable[[np.ndarray,float], np.ndarray]=model_function,p0=[2.0],refx=np.linspace(1,100,100)):\n",
    "#     # lr=LinearRegression()\n",
    "#     x_=np.array(x)/100\n",
    "#     y_=np.array(y)/100\n",
    "#     opt_paras, _ = curve_fit(model_function, x_, y_, p0=p0)\n",
    "    \n",
    "#     rmsd=((y_-model_function(x_, *opt_paras))**2).mean()\n",
    "#     # a_optimal = optimal_parameters\n",
    "#     predy = model_function(refx/100, *opt_paras)\n",
    "#     sel=(predy>=0) & (predy<=1)\n",
    "\n",
    "#     reg_x=refx[sel]\n",
    "#     reg_y=predy[sel]*100\n",
    "\n",
    "#     return (\n",
    "#         opt_paras,rmsd,\n",
    "#         reg_x,reg_y,\n",
    "    # )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def norm(x_:float):\n",
    "    # return np.sqrt(x_)\n",
    "    _pseudo_index=1.5#log(log(1.5,2),0.8)\n",
    "    _pseudo_identity=2**(x_**_pseudo_index)-1\n",
    "    return _pseudo_identity\n",
    "    \n",
    "def rev_norm(y_:float):\n",
    "    # return y_**2\n",
    "    _pseudo_index=1.5#log(log(1.5,2),0.8)\n",
    "    identity=np.log2(y_+1)**(1/_pseudo_index)\n",
    "    return identity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def get_no_ref_accession_genome(\n",
    "    domains:pd.DataFrame,ref_accession:str\n",
    ")->List[str]:\n",
    "    genomes=domains['genome_name'].unique()\n",
    "    no_ref_accession_genome=[]\n",
    "    for g in genomes:\n",
    "        if not ((domains['genome_name']==g)&(domains['domain_accession']==ref_accession)).any():\n",
    "            no_ref_accession_genome.append(g)\n",
    "    return no_ref_accession_genome\n",
    "\n",
    "def get_sequence_length(domains:pd.DataFrame,ref_accession:str,paddding=1000):\n",
    "    _=domains[domains['domain_accession']==ref_accession].copy(deep=True) # poor performance, could be slow\n",
    "    left_=_['start']-(_['hmmStart']*3)\n",
    "    right_=_['genome_length']-left_\n",
    "    left,right=left_.max(),right_.max()\n",
    "    no_ref_accession_genome=get_no_ref_accession_genome(domains,ref_accession)\n",
    "    if len(no_ref_accession_genome)>0:\n",
    "        max_unalign_length=domains[domains['genome_name'].isin(no_ref_accession_genome)]['genome_length'].max()\n",
    "    else:\n",
    "        max_unalign_length=0\n",
    "    \n",
    "    if max_unalign_length<left+right:\n",
    "        sequence_length,ref=left+right+paddding*2,left+paddding\n",
    "        return sequence_length,ref\n",
    "    else:\n",
    "        sequence_length=max_unalign_length+paddding*2\n",
    "        ref=left+paddding+(sequence_length-(left+right))/2\n",
    "    # left,right=max(0,max_unalign_length,left_.max())+paddding,max(0,max_unalign_length,right_.max())+paddding\n",
    "    # sequence_length,ref=left+right,left\n",
    "    return sequence_length,ref\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "family_dict={\n",
    "    'zika':'Flaviviridae',\n",
    "    'ebola':'Filoviridae',\n",
    "    'measles':'Paramyxoviridae'\n",
    "}\n",
    "ref_genome_dict={\n",
    "    'zika':'ZIKV||AY632535',\n",
    "    'ebola':'EBOV||AF086833',\n",
    "    'measles':'MeV||AB016162'\n",
    "}\n",
    "ref_accession_dict={\n",
    "    'zika':'PF00972',\n",
    "    'ebola':'PF00946',\n",
    "    'measles':'PF00946'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "corr_color_list = [mcolors.to_rgba(i) for i in ['darkblue','teal','orangered']] #colormaps['RdYlBu_r']#['plasma_r'] ['darkgreen','darkblue','maroon']\n",
    "\n",
    "k='ebola'\n",
    "ref_genome=ref_genome_dict[k]\n",
    "ref_accession=ref_accession_dict[k]\n",
    "suffix='synteney_'\n",
    "\n",
    "def plot_synteny_v1(k:str,ref_genome:str,ref_accession:str,\n",
    "                 corr_color_list=corr_color_list,\n",
    "                 missing_color='maroon',\n",
    "                 suffix:str='synteney'):\n",
    "    #potential parameter, no need for this stage\n",
    "    ref_palette_fetch=lambda x: x[0]\n",
    "    noalign_holder=15\n",
    "    \n",
    "    \n",
    "    missing_color_map = LinearSegmentedColormap.from_list('missing_color_map', [(1, 1, 1,1), mcolors.to_rgba(missing_color)])\n",
    "    domains=pd.read_csv(f'{k}-domains.csv')\n",
    "    sequence_length,ref=get_sequence_length(domains,ref_accession)#int(2*domains['genome_length'].max())\n",
    "    # ref=int(sequence_length/2)\n",
    "\n",
    "    identities_dict=get_identities(f'{k}_fasta')[0]\n",
    "\n",
    "    # canonical colormap\n",
    "    lr_cohort=get_lr_cohort(identities_dict,ref_genome,ref_accession=ref_accession)\n",
    "    lr_metrics={k:cal_lr(v[0],v[1]) for k,v in lr_cohort.items()}\n",
    "    # corr_color_map,situ,position_ref,metric_min,metric_max=get_corr_color_map(lr_metrics,ref_palette_fetch,corr_color_list)\n",
    "    corr_color_map,situ,position_ref,metric_min,metric_max=fix_corr_color_map(lr_metrics,ref_palette_fetch,corr_color_list)\n",
    "    # tfm=lambda x:(x-metric_min)/(metric_max-metric_min)\n",
    "    tfm=lambda x: (max(min(x,2),0.5)-0.5)/1.5\n",
    "    ref_palette={k:get_ref_palette(k,lr_metrics,corr_color_map,\n",
    "                transform=tfm,fetch=ref_palette_fetch) for k in lr_metrics.keys()}\n",
    "\n",
    "    missing_palette,max_missing=get_missing_palette(domains,ref_genome,missing_color_map)\n",
    "\n",
    "    plt.close('all')\n",
    "    sort_rdrps=parse_diamond(f'{k}_fasta/{ref_accession}-match.tsv')\n",
    "    sort_rdrps.set_index('Subject ID',inplace=True)\n",
    "    genomes=domains['genome_name'].unique()\n",
    "\n",
    "\n",
    "\n",
    "    fig,ax=plt.subplots(len(genomes)+2,1,sharex=True,\n",
    "                        figsize=(16,len(genomes)*0.3+3),\n",
    "                        height_ratios=[1.,2.]+[0.3]*len(genomes))\n",
    "    c=2\n",
    "    # fig,ax=plt.subplots(2,1,sharex=True,figsize=(30,2*2))\n",
    "    gs=gridspec.GridSpec(len(genomes)+2,2,height_ratios=[1.,2.]+[0.3]*len(genomes))\n",
    "    ax:List[Axes]\n",
    "\n",
    "    for (i,j) in sort_genome(genomes,sort_rdrps):\n",
    "        '''\n",
    "        i: genome id,\n",
    "        j: identity% to ref genome \n",
    "        '''\n",
    "        idf=domains[domains['genome_name']==i]\n",
    "        # for i,idf in domains.groupby('genome_name'):\n",
    "            # sequence_length=idf.iloc[0]['genome_length']\n",
    "        if (idf['domain_accession']==ref_accession).any():\n",
    "            #TODO better decision of genome length & del_start\n",
    "            rdrp_start=idf[idf['domain_accession']==ref_accession].iloc[0]['start']\n",
    "            rdrp_hmm_start=idf[idf['domain_accession']==ref_accession].iloc[0]['hmmStart']\n",
    "            del_start=ref-(rdrp_start-rdrp_hmm_start*3)\n",
    "        else:\n",
    "            # del_start=ref-idf.iloc[0]['genome_length']/2\n",
    "            del_start=(sequence_length-idf.iloc[0]['genome_length'])/2\n",
    "            # print(f'no rdrp warning:{i}') #TODO unified log system\n",
    "        \n",
    "        features=[]\n",
    "        for _,d in idf.iterrows():\n",
    "            strand=+1 if d['strand']=='SENSE' else -1\n",
    "            a_=d['domain_annotation'].split(':')[0]\n",
    "            if d['domain_accession'] in ref_palette:\n",
    "                identity=identities_dict[d['domain_accession']].get(i,noalign_holder)/100\n",
    "                _pseudo_identity=norm(identity)\n",
    "                color=ref_palette[d['domain_accession']](_pseudo_identity)\n",
    "            else:\n",
    "                color=missing_color_map(1.)#missing_palette[d['domain_accession']]\n",
    "            features.append(GraphicFeature(start=d['start']+del_start, end=d['end']+del_start, strand=strand, color=color,\n",
    "                                        linewidth=1, thickness=7)) #,linecolor=color\n",
    "                        # label=d['domain_accession'],fontdict={'fontsize':15}))\n",
    "        record = GraphicRecord(sequence_length=sequence_length, feature_level_height=0,features=features)\n",
    "        record.plot(ax=ax[c],figure_width=20)\n",
    "        ax[c].set_facecolor(\"white\")\n",
    "        ax[c].tick_params(axis='x', which='both', bottom=False, top=False, labelbottom=False) # remove ticks\n",
    "        ax[c].annotate(i+f'|{j}' ,(-100,0.2), xycoords='data', annotation_clip=False, ha='left',fontsize=8) \n",
    "        hide_axes(ax[c],0+del_start,idf.iloc[0]['genome_length']+del_start,sequence_length=sequence_length)\n",
    "        c+=1 \n",
    "        \n",
    "    for i in range(2):\n",
    "        ax[i].set_axis_off()\n",
    "    fig.tight_layout()\n",
    "\n",
    "    annot_ax =fig.add_subplot(gs[1,:])\n",
    "    virtual_track=[]\n",
    "    unique_domains=domains['domain_accession'].unique()\n",
    "    annot_space=np.linspace(0,sequence_length,len(unique_domains)+2)\n",
    "    for i,dm in enumerate(unique_domains):\n",
    "        annot=domains[domains['domain_accession']==dm].iloc[0]['domain_annotation'].split(':')[0]\n",
    "        label=f'{dm}:{annot}'\n",
    "        if dm in ref_palette:\n",
    "            color=ref_palette[dm](1.0)\n",
    "        else:\n",
    "            color=missing_color_map(1.)#missing_palette[dm]\n",
    "        textcolor = get_foreground(mcolors.to_hex(color))\n",
    "        virtual_track.append(GraphicFeature(start=int(annot_space[i+1]), end=int(annot_space[i+1])+1, \n",
    "                    strand=+1, color=color,label=label,\n",
    "                    linewidth=0, thickness=7, box_color=color, box_linewidth=0,\n",
    "                    label_link_color=\"white\",linecolor='white',\n",
    "                    fontdict={\"color\":textcolor,'fontsize':8}))\n",
    "    record = GraphicRecord(sequence_length=sequence_length,  features=virtual_track,feature_level_height=1) #\n",
    "    ax_, _ = record.plot(ax=annot_ax, figure_width=8,draw_line=False) \n",
    "    ax_.set_facecolor(\"white\")\n",
    "    ax_.tick_params(axis='x', which='both', bottom=False, top=False, labelbottom=False) # remove ticks\n",
    "    # genome_start_pos = rdrp_fixed_position - [x[0] for x in sars_cov_2_features if \"RdRP\" in x[3]][0]\n",
    "    # genome_end_pos = genome_start_pos + 29903\n",
    "    # ax_.plot([genome_start_pos, genome_end_pos], [0, 0], zorder=-1000, c=\"k\") \n",
    "\n",
    "    # colormap\n",
    "    cbar_ax =fig.add_subplot(gs[0,0])\n",
    "    cbar=fig.colorbar(cm.ScalarMappable(norm=Normalize(vmin=metric_min,vmax=metric_max), cmap=corr_color_map),cax=cbar_ax,orientation='horizontal')\n",
    "    cbar_tick_locations = [metric_min,1,metric_max]#np.linspace(0,1,4)\n",
    "    cbar.set_ticks(cbar_tick_locations)\n",
    "    cbar.set_ticklabels([f'{i:.1f}' for i in cbar_tick_locations],fontsize=8)\n",
    "    cbar.set_label('Lr `k` with ref domain',fontsize=12,loc='center')\n",
    "\n",
    "    cbar_ax1 =fig.add_subplot(gs[0,1])\n",
    "    cbar1=fig.colorbar(cm.ScalarMappable(norm=FuncNorm(functions=(norm,rev_norm),vmin=0.,vmax=1.), \n",
    "                                        cmap=ref_palette[ref_accession]),cax=cbar_ax1,orientation='horizontal')\n",
    "    cbar_tick_locations = np.array([0,0.5,0.8,1])\n",
    "    cbar1.set_ticks(cbar_tick_locations)\n",
    "    cbar1.set_ticklabels([f'{i:.1f}' for i in cbar_tick_locations],fontsize=8)\n",
    "    cbar1.set_label('Identity to RefGenome',fontsize=12,loc='center')\n",
    "\n",
    "    # cbar_ax2 =fig.add_subplot(gs[0,2])\n",
    "    # cbar2=fig.colorbar(cm.ScalarMappable(norm=Normalize(vmin=0.,vmax=max_missing), cmap=missing_color_map),cax=cbar_ax2,orientation='horizontal')\n",
    "    # cbar_tick_locations = np.array(np.linspace(0,max_missing,5))\n",
    "    # cbar2.set_ticks(cbar_tick_locations)\n",
    "    # cbar2.set_ticklabels([f'{i:.0f}' for i in cbar_tick_locations],fontsize=8)\n",
    "    # cbar2.set_label('Missing Gene Freq',fontsize=12,loc='center')\n",
    "\n",
    "    # fig.set_dpi(20)\n",
    "    fig.savefig(f'{k}-{suffix}.pdf')\n",
    "    # plt.show()\n",
    "    plt.close('all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1.0, 0.0, 0.0, 1.0)"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mcolors.to_rgba('red')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "k='zika'\n",
    "ref_genome=ref_genome_dict[k]\n",
    "ref_accession=ref_accession_dict[k]\n",
    "suffix='synteney_1'\n",
    "for k in ref_genome_dict:\n",
    "    plot_synteny_v1(k,ref_genome_dict[k],ref_accession_dict[k],\n",
    "                    corr_color_list=['green','black','red'],missing_color='purple',\n",
    "                    suffix=suffix)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rdrp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
