{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "post processing of hhblits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle as pkl\n",
    "from typing import List,Dict,Tuple,Union\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "search_result:List[Tuple[Path,List[Tuple[Path,List[str]]]]]=pkl.load(open('blit_out.pkl','rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def proindice_to_genindice(genome:Union[str,Seq],prob:int,proe:int,transannot:str)->Tuple[int,int]:\n",
    "    frame,direction=transannot\n",
    "    frame=int(frame)\n",
    "    b_,e_=frame+prob*3,frame+proe*3\n",
    "    if direction=='s':\n",
    "        return b_,e_\n",
    "    else:\n",
    "        return len(genome)-e_, len(genome)-b_\n",
    "\n",
    "def genindice_to_proindice(genome:Union[str,Seq],genb:int,gene:int,transannot:str)->Tuple[int,int]:\n",
    "    frame,direction=transannot\n",
    "    frame=int(frame)\n",
    "    \n",
    "    if direction=='s':\n",
    "        b_,e_=(genb-frame)//3,(gene-frame)//3\n",
    "        return b_,e_\n",
    "    else:\n",
    "        return (len(genome)-frame-gene)//3,(len(genome)-frame-genb)//3\n",
    "        \n",
    "\n",
    "def genome_annot_indice(genome:Union[str,Seq],genb:int,gene:int,transannot:str)->Tuple[int,int]:\n",
    "    'return: genseq segments '\n",
    "    frame,direction=transannot\n",
    "    # frame=int(frame)\n",
    "    o=genome[genb:gene]\n",
    "    o=o if direction=='s' else o[::-1]\n",
    "    return o"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dna_features_viewer import GraphicFeature, GraphicRecord\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as clr\n",
    "import matplotlib.patches as patches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from collections import namedtuple\n",
    "from matplotlib import colormaps\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "Hit=namedtuple('Hit',('domain_accession','domain_annotation',\n",
    "    'prob','eval','pval','score','ss','begin','end','hmmbegin','hmmend','hmmlength'))\n",
    "SegMeta=namedtuple('SegMeta',(\"transannot\",'seg_b','seg_e'))\n",
    "\n",
    "def split_hit_line(hit_line:str)->List[str]:\n",
    "    i=hit_line\n",
    "    assert len(i)>94\n",
    "    return [i[:4],i[4:34],i[34:40],i[40:48],i[48:56],i[56:63],\n",
    "            i[63:69],i[69:74],i[74:85],i[85:94],i[94:]]\n",
    "\n",
    "def parse_splited_hit_line(hit:List[str])->Hit:\n",
    "    domain_accession,*_=hit[1].split(';')\n",
    "    domain_accession=domain_accession.split('.')[0].strip()\n",
    "    if len(_)>0:\n",
    "        domain_annotation=_[0].strip()\n",
    "    else:\n",
    "        domain_annotation=domain_accession\n",
    "        \n",
    "    prob=float(hit[2].strip())\n",
    "    e=float(hit[3].strip())\n",
    "    p=float(hit[4].strip())\n",
    "    score=float(hit[5].strip())\n",
    "    ss=float(hit[6].strip())\n",
    "    begin,end=[int(i.strip())-1 for i in hit[8].split('-')]\n",
    "    hmmbegin,hmmend=[int(i.strip())-1 for i in hit[9].split('-')]\n",
    "    hmmlength=int(hit[10].strip().strip('(').strip(')'))\n",
    "    h=Hit(domain_accession=domain_accession,domain_annotation=domain_annotation,\n",
    "        prob=prob,eval=e,pval=p,score=score,ss=ss,\n",
    "        begin=begin,end=end,hmmbegin=hmmbegin,hmmend=hmmend,hmmlength=hmmlength)\n",
    "    return h\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import accumulate\n",
    "def get_color(genome_position:float,prob:float):\n",
    "    ''' domain colors based on relative position'''\n",
    "    cm=mcolors.LinearSegmentedColormap.from_list('tmp',[(1, 1, 1), colormaps['rainbow'](genome_position)], N=512)\n",
    "    return cm(prob/100)\n",
    "\n",
    "def get_color_dict(clan_annotation:pd.Series,used_cm=['tab10','Set2','Accent','Set3','Pastel2','Dark2']):\n",
    "    ''' clan colors '''\n",
    "    uni_annotation=clan_annotation.unique()\n",
    "    lengths=[colormaps[i].N for i in used_cm]\n",
    "    assert len(uni_annotation)<=sum(lengths),(f'no enough colors:'\n",
    "         f'{len(uni_annotation)} vs {sum(lengths)}')\n",
    "    out_dict={}\n",
    "    cpkt=[0]+list(accumulate(lengths))\n",
    "    cm_pointer=0\n",
    "    for i,j in enumerate(uni_annotation):\n",
    "        if i>=cpkt[cm_pointer+1]:\n",
    "            cm_pointer+=1\n",
    "        color=colormaps[used_cm[cm_pointer]](i-cpkt[cm_pointer])\n",
    "        out_dict[j]=color\n",
    "    return out_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from matplotlib.backends.backend_pdf import PdfPages\n",
    "# if 0:\n",
    "#     with PdfPages('poc2.pdf') as pdf:\n",
    "#         for genome_res in search_result[:2]:\n",
    "#             genome_res:Tuple[Path,List[Tuple[Path,List[str]]]]\n",
    "#             genome,o=parse_genome_res(genome_res)\n",
    "#             fig,ax=plt.subplots(1,1,sharex=True,figsize=(30,10))\n",
    "#             features=[]\n",
    "#             len_genome=len(genome)\n",
    "#             for (transannot,prob,proe),hit in o:\n",
    "#                 hit:Hit\n",
    "                \n",
    "#                 genomeb,genomee=proindice_to_genindice(genome,prob+hit.begin,prob+hit.end,transannot)\n",
    "#                 strand=1 if transannot.endswith('s') else -1\n",
    "#                 rele_pos=(genomeb+genomee)/2/len_genome\n",
    "#                 features.append(GraphicFeature(start=genomeb, end=genomee, strand=strand,\n",
    "#                                 label=hit.domain_annotation,color=get_color(rele_pos,hit.prob)))\n",
    "                \n",
    "#             record = GraphicRecord(sequence_length=len_genome, features=features)\n",
    "#             record.plot(ax=ax,figure_width=20)\n",
    "#             fig.suptitle(genome.id)\n",
    "#             pdf.savefig(fig)\n",
    "#             plt.close(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "involve clans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from functools import partial\n",
    "\n",
    "def df_to_dict(df:pd.DataFrame,k:str,v:str):\n",
    "    return df.set_index(k)[v].T.to_dict()\n",
    "clans=pd.read_csv('Pfam-A.clans.tsv',delimiter='\\t',header=None)\n",
    "clans.columns=['domain','clan','clan_name','domain_name','domain_annotation']\n",
    "clan_dict=df_to_dict(clans,'domain','clan')\n",
    "clan_name_dict=df_to_dict(clans,'domain','clan_name')\n",
    "\n",
    "def query_clan(domain_accession:str,domain_annotation):\n",
    "    clan_accession=clan_dict.get(domain_accession,np.nan)\n",
    "    if clan_accession!=clan_accession:\n",
    "        clan_accession=domain_accession\n",
    "    clan_annotation=clan_name_dict.get(domain_accession,np.nan)\n",
    "    if clan_annotation!=clan_annotation:\n",
    "        clan_annotation=domain_annotation\n",
    "    return clan_accession,clan_annotation\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "# genome,o=parse_genome_res(genome_res)\n",
    "def parse_genome_res(genome_res:Tuple[Path,List[Tuple[Path,List[str]]]]):\n",
    "    stem=genome_res[0].stem\n",
    "    stems=stem.replace(':segs','').split('|')\n",
    "    genome_fasta_path=genome_res[0].with_stem(stem.replace('segs','genome'))\n",
    "    genome:SeqRecord=SeqIO.read(genome_fasta_path,'fasta')\n",
    "    o:List[Tuple[SegMeta,Hit]]=[]\n",
    "    \n",
    "    for seg_res in genome_res[1]:\n",
    "        if len(seg_res[1])>0:\n",
    "            _,transannot,prob,proe=seg_res[0].stem.split('#')\n",
    "            prob,proe=int(prob),int(proe)\n",
    "            segmeta=SegMeta(transannot=transannot,seg_b=prob,seg_e=proe)\n",
    "            for h in seg_res[1]:\n",
    "                hit=parse_splited_hit_line(h)\n",
    "                o.append((segmeta,hit))   \n",
    "    return genome,o\n",
    "def genome_res_to_hits_df(genome:SeqRecord,parse_o:List[Tuple[SegMeta,Hit]])->pd.DataFrame:\n",
    "    hits_df=pd.concat([pd.DataFrame([i[1] for i in parse_o]),pd.DataFrame([i[0] for i in parse_o])],axis=1) \n",
    "    hits_df['gen_b'],hits_df['gen_e']=np.vectorize(partial(\n",
    "        proindice_to_genindice,genome=genome))(\n",
    "        prob=hits_df['begin']+hits_df['seg_b'],\n",
    "        proe=hits_df['end']+hits_df['seg_b'],\n",
    "        transannot=hits_df['transannot'])\n",
    "\n",
    "    hits_df['gen_hmm_b'],hits_df['gen_hmm_e']=np.vectorize(partial(\n",
    "        proindice_to_genindice,genome=genome))(\n",
    "        prob=hits_df['begin']-hits_df['hmmbegin']+1+hits_df['seg_b'],\n",
    "        proe=hits_df['end']+hits_df['hmmlength']-hits_df['hmmend']+hits_df['seg_b'],\n",
    "        transannot=hits_df['transannot'])\n",
    "    hits_df['clan_accession'],hits_df['clan_annotation']=np.vectorize(query_clan)(\n",
    "        hits_df['domain_accession'],hits_df['domain_annotation'])\n",
    "    return hits_df\n",
    "\n",
    "\n",
    "def inspect_genome_res(genome_res:Tuple[Path,List[Tuple[Path,List[str]]]]):\n",
    "    genome,o=parse_genome_res(genome_res)\n",
    "    hits_df=genome_res_to_hits_df(genome,o)\n",
    "    genome_id=genome.id\n",
    "    count=0\n",
    "    len_genome=len(genome)\n",
    "    color_dict=get_color_dict(hits_df['clan_annotation'])\n",
    "    with PdfPages(f'{genome_id}.pdf') as pdf:\n",
    "        while count<4:\n",
    "            fig,ax=plt.subplots(1,1,sharex=True,figsize=(30,12))\n",
    "            features=[]\n",
    "            titles=[f'{genome_id}-origin\\n50prob-domaincolor-hitindice',f'{genome_id}-filterdomain\\n90prob-domaincolor-hitindice',\n",
    "                    f'{genome_id}-color-as-clan\\n90prob-clancolor-hitindice',f'{genome_id}-cover-whole-hmm\\n90prob-clancolor-hmmindice']\n",
    "            for _,s in hits_df.iterrows():\n",
    "                strand=1 if s['transannot'].endswith('s') else -1\n",
    "                if count==0:\n",
    "                    if s['prob']>=0:\n",
    "                        genomeb,genomee=s['gen_b'],s['gen_e']\n",
    "                        rele_pos=(genomeb+genomee)/2/len_genome\n",
    "                        features.append(GraphicFeature(start=genomeb, end=genomee, strand=strand,\n",
    "                                        label=s['domain_annotation'],color=get_color(rele_pos,s['prob'])))\n",
    "                elif count==1:\n",
    "                    if s['prob']>=90:\n",
    "                        genomeb,genomee=s['gen_b'],s['gen_e']\n",
    "                        rele_pos=(genomeb+genomee)/2/len_genome\n",
    "                        features.append(GraphicFeature(start=genomeb, end=genomee, strand=strand,\n",
    "                                        label=s['domain_annotation'],color=get_color(rele_pos,s['prob'])))\n",
    "                elif count==2:\n",
    "                    if s['prob']>=90:\n",
    "                        genomeb,genomee=s['gen_b'],s['gen_e']\n",
    "                        features.append(GraphicFeature(start=genomeb, end=genomee, strand=strand,\n",
    "                                        label=s['clan_annotation'],color=color_dict[s['clan_annotation']]))\n",
    "                elif count==3:\n",
    "                    if s['prob']>=90:\n",
    "                        genomeb,genomee=s['gen_hmm_b'],s['gen_hmm_e']\n",
    "                        features.append(GraphicFeature(start=genomeb, end=genomee, strand=strand,\n",
    "                                        label=s['clan_annotation'],color=color_dict[s['clan_annotation']]))\n",
    "                    \n",
    "            record = GraphicRecord(sequence_length=len_genome, features=features)\n",
    "            record.plot(ax=ax,figure_width=20)\n",
    "            # fig.show()\n",
    "            fig.suptitle(titles[count],fontsize=20)\n",
    "            pdf.savefig(fig)\n",
    "            plt.close(fig)\n",
    "            count=count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from matplotlib import rcParams\n",
    "# type(rcParams['axes.prop_cycle'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def all_in_one_parse(genome_res:Tuple[Path,List[Tuple[Path,List[str]]]]):\n",
    "    genome,parse_o=parse_genome_res(genome_res)\n",
    "    hits_df=genome_res_to_hits_df(genome,parse_o)\n",
    "    return genome,parse_o,hits_df\n",
    "\n",
    "genome_res = search_result[1]\n",
    "genome,parse_o,hits_df=all_in_one_parse(genome_res)\n",
    "# genome,o=parse_genome_res(genome_res)\n",
    "# hits_df=genome_res_to_hits_df(genome,o)\n",
    "\n",
    "# inspect_genome_res(genome_res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "search_result_dict = {i[0].stem.strip(':segs'):i for i in search_result}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "zika_res=search_result_dict['ZIKV||AY632535']\n",
    "zika_os=all_in_one_parse(zika_res)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/hugheslab1/zfdeng/.local/lib/python3.12/site-packages/numpy/lib/function_base.py:2455: RuntimeWarning: invalid value encountered in query_clan (vectorized)\n",
      "  outputs = ufunc(*inputs)\n",
      "/home/hugheslab1/zfdeng/.local/lib/python3.12/site-packages/numpy/lib/function_base.py:2455: RuntimeWarning: invalid value encountered in query_clan (vectorized)\n",
      "  outputs = ufunc(*inputs)\n",
      "/home/hugheslab1/zfdeng/.local/lib/python3.12/site-packages/numpy/lib/function_base.py:2455: RuntimeWarning: invalid value encountered in query_clan (vectorized)\n",
      "  outputs = ufunc(*inputs)\n",
      "/home/hugheslab1/zfdeng/.local/lib/python3.12/site-packages/numpy/lib/function_base.py:2455: RuntimeWarning: invalid value encountered in query_clan (vectorized)\n",
      "  outputs = ufunc(*inputs)\n",
      "/home/hugheslab1/zfdeng/.local/lib/python3.12/site-packages/numpy/lib/function_base.py:2455: RuntimeWarning: invalid value encountered in query_clan (vectorized)\n",
      "  outputs = ufunc(*inputs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WPSSPV||MG60005\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/hugheslab1/zfdeng/.local/lib/python3.12/site-packages/numpy/lib/function_base.py:2455: RuntimeWarning: invalid value encountered in query_clan (vectorized)\n",
      "  outputs = ufunc(*inputs)\n"
     ]
    }
   ],
   "source": [
    "for i in ['ZIKV||AY632535','EBOV||AF086833',\n",
    "    'MeV||AB016162','SARS-CoV-2||MN908947',\n",
    "    'RPbV||KC815311','TBEV-FE||X07755',\n",
    "    'WPSSPV||MG600057','KNDV||MW093492']:\n",
    "    try:\n",
    "        genome_res=search_result_dict[i]\n",
    "        inspect_genome_res(genome_res)\n",
    "    except:\n",
    "        print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_res=search_result_dict['OSV-1||MF324848']\n",
    "inspect_genome_res(genome_res)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "'ZIKV||AY632535'\n",
    "    hint: too brutal to use hmm indice directly\n",
    "    cause pseudo overlap in Orbi_VP1,P-loop_NTPase\n",
    "    some families are actually multidomain (e.g. RdRP3); partial hit is ONLY partial hit.\n",
    "\n",
    "'EBOV||AF086833'\n",
    "    pseudo double hit of Filo_glycop\n",
    "    G-7-MTase & NADP_Rossmann(e.g Fstj): mighe be the same thing samething\n",
    "\n",
    "\n",
    "'MeV||AB016162'\n",
    "    pseudo double hit of Paramyxovirin_C\n",
    "\n",
    "'SARS-CoV-2||MN908947'\n",
    "    so many hits, disaster\n",
    "    to be analysed again after hhm/hhm-align and redefine clan\n",
    "    \n",
    "\n",
    "'RPbV||KC815311':\n",
    "    homology of zika, classic false-negative with PF00998 instead of zika's rdrp (PF00792) hit\n",
    "    one more hit found (1st on the left)\n",
    "\n",
    "'TBEV-FE||X07755'\n",
    "    homology of zika, out-of-distribution\n",
    "    very tidy, perfect\n",
    "\n",
    "'WPSSPV||MG600057'\n",
    "    homology of measles, with long segs of false-negative (no hit in Interproscan)\n",
    "    no more domain hit\n",
    "\n",
    "'KNDV||MW093492\n",
    "    homology of ebola, with long segs of false-negative (no hit in Interproscan)\n",
    "    some DUFs and ZFs hit in that missing area, with low(50-90) prob\n",
    "\n",
    "'OSV-1||MF324848'\n",
    "    homology of covid, random chosen."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Legend for each cases:\n",
    "1. Standard refs for Covid/Zika/Ebola/Measles\n",
    "    ZIKV||AY632535\n",
    "\n",
    "    EBOV||AF086833\n",
    "\n",
    "    MeV||AB016162\n",
    "\n",
    "    SARS-CoV-2||MN908947\n",
    "\n",
    "2. Representative homologs:\n",
    "    RPbV||KC815311: \n",
    "        homology of zika, classic false-negative with PF00998 instead of zika's rdrp (PF00792) hit\n",
    "\n",
    "    TBEV-FE||X07755:\n",
    "        homology of zika, out-of-distribution orphan\n",
    "\n",
    "    WPSSPV||MG600057\n",
    "        homology of measles, with long segs of false-negative (no hit in Interproscan)\n",
    "\n",
    "    KNDV||MW093492\n",
    "        homology of ebola, with long segs of false-negative (no hit in Interproscan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from glob import glob\n",
    "for i in Path('.').iterdir():\n",
    "    if i.suffix=='.pdf':\n",
    "        new_stem=i.stem.replace('|','#')\n",
    "        i.rename(i.with_stem(new_stem))\n",
    "    \n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Progress:\n",
    "hh-blits returns every possible hits as we expected\n",
    "\n",
    "hh-blits could hit domains ignored by Interproscan (RPbV||KC815311,OSV-1||MF324848)\n",
    "\n",
    "hh-blits could fix false loops (multi-hit. e.g MeV||AB016162)\n",
    "\n",
    "Issues:\n",
    "hh-blits is slower than interproscan (avg: 20core * 20s; interproscan: 4core * 15s) \n",
    "    reason: interproscan could predict ORFs, \n",
    "            while translation of multiple segs (most of them are useless) will be analyzed tandemly in hh-blits\n",
    "\n",
    "False ORF could yield false hit (small fragments) with high probability\n",
    "\n",
    "Pfam clan is very rough. rerun with hhblits is definitely needed.\n",
    "\n",
    "Too brutal to use hmm indice directly, which might cause pseudo overlap\n",
    "\n",
    "hitting of domains ignored by Interproscan is still not perfect (WPSSPV||MG600057, KNDV||MW093492)\n",
    "\n",
    "SARS-CoV-2||MN908947: too many hits to be disaster\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>domain_accession</th>\n",
       "      <th>domain_annotation</th>\n",
       "      <th>prob</th>\n",
       "      <th>eval</th>\n",
       "      <th>pval</th>\n",
       "      <th>score</th>\n",
       "      <th>ss</th>\n",
       "      <th>begin</th>\n",
       "      <th>end</th>\n",
       "      <th>hmmbegin</th>\n",
       "      <th>...</th>\n",
       "      <th>hmmlength</th>\n",
       "      <th>transannot</th>\n",
       "      <th>seg_b</th>\n",
       "      <th>seg_e</th>\n",
       "      <th>gen_b</th>\n",
       "      <th>gen_e</th>\n",
       "      <th>gen_hmm_b</th>\n",
       "      <th>gen_hmm_e</th>\n",
       "      <th>clan_accession</th>\n",
       "      <th>clan_annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PF19215</td>\n",
       "      <td>CoV_NSP15_C</td>\n",
       "      <td>100.0</td>\n",
       "      <td>3.700000e-36</td>\n",
       "      <td>7.200000e-40</td>\n",
       "      <td>303.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1174</td>\n",
       "      <td>1294</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>155</td>\n",
       "      <td>0s</td>\n",
       "      <td>2069</td>\n",
       "      <td>3485</td>\n",
       "      <td>9729</td>\n",
       "      <td>10089</td>\n",
       "      <td>9729</td>\n",
       "      <td>10092</td>\n",
       "      <td>CL0695</td>\n",
       "      <td>EndoU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PF17977</td>\n",
       "      <td>zf-RING_13</td>\n",
       "      <td>99.7</td>\n",
       "      <td>1.800000e-22</td>\n",
       "      <td>3.300000e-26</td>\n",
       "      <td>174.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>636</td>\n",
       "      <td>696</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>68</td>\n",
       "      <td>0s</td>\n",
       "      <td>2069</td>\n",
       "      <td>3485</td>\n",
       "      <td>8115</td>\n",
       "      <td>8295</td>\n",
       "      <td>8115</td>\n",
       "      <td>8304</td>\n",
       "      <td>CL0229</td>\n",
       "      <td>RING</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PF00978</td>\n",
       "      <td>RdRP_2</td>\n",
       "      <td>99.6</td>\n",
       "      <td>1.900000e-21</td>\n",
       "      <td>4.400000e-25</td>\n",
       "      <td>218.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>331</td>\n",
       "      <td>571</td>\n",
       "      <td>148</td>\n",
       "      <td>...</td>\n",
       "      <td>440</td>\n",
       "      <td>0s</td>\n",
       "      <td>2069</td>\n",
       "      <td>3485</td>\n",
       "      <td>7200</td>\n",
       "      <td>7920</td>\n",
       "      <td>6759</td>\n",
       "      <td>8109</td>\n",
       "      <td>CL0027</td>\n",
       "      <td>RdRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PF00998</td>\n",
       "      <td>RdRP_3</td>\n",
       "      <td>99.5</td>\n",
       "      <td>6.300000e-19</td>\n",
       "      <td>1.500000e-22</td>\n",
       "      <td>197.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>331</td>\n",
       "      <td>574</td>\n",
       "      <td>135</td>\n",
       "      <td>...</td>\n",
       "      <td>479</td>\n",
       "      <td>0s</td>\n",
       "      <td>2069</td>\n",
       "      <td>3485</td>\n",
       "      <td>7200</td>\n",
       "      <td>7929</td>\n",
       "      <td>6798</td>\n",
       "      <td>8238</td>\n",
       "      <td>CL0027</td>\n",
       "      <td>RdRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PF17873</td>\n",
       "      <td>Rep_1B</td>\n",
       "      <td>99.4</td>\n",
       "      <td>4.700000e-18</td>\n",
       "      <td>8.600000e-22</td>\n",
       "      <td>140.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>709</td>\n",
       "      <td>760</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>53</td>\n",
       "      <td>0s</td>\n",
       "      <td>2069</td>\n",
       "      <td>3485</td>\n",
       "      <td>8334</td>\n",
       "      <td>8487</td>\n",
       "      <td>8334</td>\n",
       "      <td>8490</td>\n",
       "      <td>CL0105</td>\n",
       "      <td>Hybrid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>PF12046</td>\n",
       "      <td>CCB1</td>\n",
       "      <td>74.1</td>\n",
       "      <td>1.300000e-01</td>\n",
       "      <td>2.800000e-05</td>\n",
       "      <td>35.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86</td>\n",
       "      <td>132</td>\n",
       "      <td>62</td>\n",
       "      <td>...</td>\n",
       "      <td>171</td>\n",
       "      <td>2s</td>\n",
       "      <td>792</td>\n",
       "      <td>977</td>\n",
       "      <td>2636</td>\n",
       "      <td>2774</td>\n",
       "      <td>2453</td>\n",
       "      <td>2948</td>\n",
       "      <td>PF12046</td>\n",
       "      <td>CCB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>PF02340</td>\n",
       "      <td>PRRSV_Env</td>\n",
       "      <td>77.3</td>\n",
       "      <td>9.200000e-02</td>\n",
       "      <td>2.000000e-05</td>\n",
       "      <td>40.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>211</td>\n",
       "      <td>286</td>\n",
       "      <td>139</td>\n",
       "      <td>...</td>\n",
       "      <td>233</td>\n",
       "      <td>2s</td>\n",
       "      <td>3492</td>\n",
       "      <td>3792</td>\n",
       "      <td>11111</td>\n",
       "      <td>11336</td>\n",
       "      <td>10697</td>\n",
       "      <td>11384</td>\n",
       "      <td>PF02340</td>\n",
       "      <td>PRRSV_Env</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>PF00951</td>\n",
       "      <td>Arteri_Gl</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.800000e-65</td>\n",
       "      <td>3.500000e-69</td>\n",
       "      <td>432.9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>77</td>\n",
       "      <td>244</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>173</td>\n",
       "      <td>2s</td>\n",
       "      <td>4012</td>\n",
       "      <td>4257</td>\n",
       "      <td>12269</td>\n",
       "      <td>12770</td>\n",
       "      <td>12254</td>\n",
       "      <td>12773</td>\n",
       "      <td>PF00951</td>\n",
       "      <td>Arteri_Gl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>PF01481</td>\n",
       "      <td>Arteri_nucleo</td>\n",
       "      <td>100.0</td>\n",
       "      <td>5.800000e-46</td>\n",
       "      <td>1.100000e-49</td>\n",
       "      <td>271.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19</td>\n",
       "      <td>123</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>115</td>\n",
       "      <td>2s</td>\n",
       "      <td>4423</td>\n",
       "      <td>4549</td>\n",
       "      <td>13328</td>\n",
       "      <td>13640</td>\n",
       "      <td>13301</td>\n",
       "      <td>13643</td>\n",
       "      <td>PF01481</td>\n",
       "      <td>Arteri_nucleo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>PF16623</td>\n",
       "      <td>WW_FCH_linker</td>\n",
       "      <td>67.3</td>\n",
       "      <td>2.200000e-01</td>\n",
       "      <td>5.100000e-05</td>\n",
       "      <td>27.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>81</td>\n",
       "      <td>89</td>\n",
       "      <td>66</td>\n",
       "      <td>...</td>\n",
       "      <td>93</td>\n",
       "      <td>2a</td>\n",
       "      <td>1659</td>\n",
       "      <td>1762</td>\n",
       "      <td>8520</td>\n",
       "      <td>8544</td>\n",
       "      <td>8463</td>\n",
       "      <td>8739</td>\n",
       "      <td>PF16623</td>\n",
       "      <td>WW_FCH_linker</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>102 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    domain_accession domain_annotation   prob          eval          pval  \\\n",
       "0            PF19215       CoV_NSP15_C  100.0  3.700000e-36  7.200000e-40   \n",
       "1            PF17977        zf-RING_13   99.7  1.800000e-22  3.300000e-26   \n",
       "2            PF00978            RdRP_2   99.6  1.900000e-21  4.400000e-25   \n",
       "3            PF00998            RdRP_3   99.5  6.300000e-19  1.500000e-22   \n",
       "4            PF17873            Rep_1B   99.4  4.700000e-18  8.600000e-22   \n",
       "..               ...               ...    ...           ...           ...   \n",
       "97           PF12046              CCB1   74.1  1.300000e-01  2.800000e-05   \n",
       "98           PF02340         PRRSV_Env   77.3  9.200000e-02  2.000000e-05   \n",
       "99           PF00951         Arteri_Gl  100.0  1.800000e-65  3.500000e-69   \n",
       "100          PF01481     Arteri_nucleo  100.0  5.800000e-46  1.100000e-49   \n",
       "101          PF16623     WW_FCH_linker   67.3  2.200000e-01  5.100000e-05   \n",
       "\n",
       "     score   ss  begin   end  hmmbegin  ...  hmmlength  transannot seg_b  \\\n",
       "0    303.1  0.0   1174  1294         1  ...        155          0s  2069   \n",
       "1    174.0  0.0    636   696         1  ...         68          0s  2069   \n",
       "2    218.1  0.0    331   571       148  ...        440          0s  2069   \n",
       "3    197.8  0.0    331   574       135  ...        479          0s  2069   \n",
       "4    140.5  0.0    709   760         1  ...         53          0s  2069   \n",
       "..     ...  ...    ...   ...       ...  ...        ...         ...   ...   \n",
       "97    35.7  0.0     86   132        62  ...        171          2s   792   \n",
       "98    40.3  0.0    211   286       139  ...        233          2s  3492   \n",
       "99   432.9  0.0     77   244         6  ...        173          2s  4012   \n",
       "100  271.1  0.0     19   123        10  ...        115          2s  4423   \n",
       "101   27.5  0.0     81    89        66  ...         93          2a  1659   \n",
       "\n",
       "     seg_e  gen_b  gen_e  gen_hmm_b  gen_hmm_e  clan_accession clan_annotation  \n",
       "0     3485   9729  10089       9729      10092          CL0695           EndoU  \n",
       "1     3485   8115   8295       8115       8304          CL0229            RING  \n",
       "2     3485   7200   7920       6759       8109          CL0027            RdRP  \n",
       "3     3485   7200   7929       6798       8238          CL0027            RdRP  \n",
       "4     3485   8334   8487       8334       8490          CL0105          Hybrid  \n",
       "..     ...    ...    ...        ...        ...             ...             ...  \n",
       "97     977   2636   2774       2453       2948         PF12046            CCB1  \n",
       "98    3792  11111  11336      10697      11384         PF02340       PRRSV_Env  \n",
       "99    4257  12269  12770      12254      12773         PF00951       Arteri_Gl  \n",
       "100   4549  13328  13640      13301      13643         PF01481   Arteri_nucleo  \n",
       "101   1762   8520   8544       8463       8739         PF16623   WW_FCH_linker  \n",
       "\n",
       "[102 rows x 21 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zika_os[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AAbV||GBBW01007738:segs',\n",
       " 'OSV-1||MF324848:segs',\n",
       " 'EAV||X53459:segs',\n",
       " 'APRAV||KP026921:segs',\n",
       " 'SHFV||AF180391:segs',\n",
       " 'SHEV||KM677927:segs',\n",
       " 'FSVV||KR862307:segs',\n",
       " 'ZMbV-1||KT166441:segs',\n",
       " 'KRCV-2||KC787658:segs',\n",
       " 'DeBMAV||KP126831:segs',\n",
       " 'KRTGV-1||JX473849:segs',\n",
       " 'PBJV||KR139839:segs',\n",
       " 'KKCBV||KT447550:segs',\n",
       " 'MYBV-1||KM110938:segs',\n",
       " 'KRCV-1||KC787630:segs',\n",
       " 'PRRSV-2||U87392:segs',\n",
       " 'RtMrufAV||KP280006:segs',\n",
       " 'RtClonAV||KU302440:segs',\n",
       " 'RtEiAV||KY369968:segs',\n",
       " 'PRRSV-1||M96262:segs',\n",
       " 'RtMcAV||KY369967:segs',\n",
       " 'LDV||U15146:segs',\n",
       " 'RtClanAV||KY369969:segs',\n",
       " 'WPDV||JN116253:segs',\n",
       " 'TSHSV-NX1||MH447987:segs',\n",
       " 'CBHPTAV||MG600025:segs',\n",
       " 'GGGSAV||MG600023:segs',\n",
       " 'HOFAV||MG600022:segs',\n",
       " 'MLeV||GECV01031551:segs',\n",
       " 'BtCoV-AMA-L-F||MT663548:segs',\n",
       " 'BtCoV CDPHE15||KF430219:segs',\n",
       " 'HipPBCoV-CHB25||MN611525:segs',\n",
       " 'ACoV-WA3607||MK472070:segs',\n",
       " 'BtCoV HKU10||JQ989270:segs',\n",
       " 'BtRf-AlphaCoV||KJ473807:segs',\n",
       " 'HCoV_229E||AF304460:segs',\n",
       " 'HCoV_229E||JX503061:segs',\n",
       " 'LRNV||KF294380:segs',\n",
       " 'MCoV||HM245925:segs',\n",
       " 'Mi-BatCoV_1A||EU420138:segs',\n",
       " 'Mi-BatCoV_HKU8||EU420139:segs',\n",
       " 'BtMr-AlphaCoV||KJ473806:segs',\n",
       " 'TyBCoV-HKU33||MK720944:segs',\n",
       " 'ACoV-WA2028||MK472068:segs',\n",
       " 'BtNv-AlphaCoV||KJ473809:segs',\n",
       " 'PK-BatCoV||MH938449:segs',\n",
       " 'ACoV-WA1087||MK472067:segs',\n",
       " 'PEDV||AF353511:segs',\n",
       " 'Sc-BatCoV_512||DQ648858:segs',\n",
       " 'Rh-BatCoV_HKU2||EF203064:segs',\n",
       " 'HCoV_NL63||AY567487:segs',\n",
       " 'BtKYNL63||KY073745:segs',\n",
       " 'Sa-CoV_T14||KY370053:segs',\n",
       " 'Sm-CoV_X74||KY967715:segs',\n",
       " 'TGEV||AJ271965:segs',\n",
       " 'HCoV_OC43||AY585228:segs',\n",
       " 'ChRCoV_HKU24||KM349742:segs',\n",
       " 'HCoV_HKU1||AY597011:segs',\n",
       " 'MHV||AY700211:segs',\n",
       " 'MrufCoV_2JL14||KY370046:segs',\n",
       " 'Bat_Hp-BetaCoV||KF636752:segs',\n",
       " 'EriCoV||KC545383:segs',\n",
       " 'MERS-CoV||JX869059:segs',\n",
       " 'Pi-BatCoV_HKU5||EF065509:segs',\n",
       " 'Ty-BatCoV-HKU4||EF065505:segs',\n",
       " 'Ei-BatCoV_C704||MG693168:segs',\n",
       " 'Ro-BatCoV_GCCDC1||KU762338:segs',\n",
       " 'Ro-BatCoV_HKU9||EF065513:segs',\n",
       " 'SARS-CoV||AY274119:segs',\n",
       " 'SARSr-CoV||KY352407:segs',\n",
       " 'SARS-CoV-2||MN908947:segs',\n",
       " 'SARS-CoV||AY613950:segs',\n",
       " 'WiCoV_HKU20||JQ065048:segs',\n",
       " 'BulCV_HKU11||FJ376619:segs',\n",
       " 'CMCoV_HKU21||JQ065049:segs',\n",
       " 'PoCoV_HKU15||JQ065043:segs',\n",
       " 'MunCV_HKU13||FJ376622:segs',\n",
       " 'WECoV_HKU16||JQ065044:segs',\n",
       " 'NHCoV_HKU19||JQ065047:segs',\n",
       " 'BcanCoV_CB17||MK359255:segs',\n",
       " 'BWCoV||EU111742:segs',\n",
       " 'BdCoV_HKU22||KF793824:segs',\n",
       " 'IBV||M95169:segs',\n",
       " 'ACoV_9203||KR902510:segs',\n",
       " 'DuCoV_2714||KM454473:segs',\n",
       " 'PsNV||MK611985:segs',\n",
       " 'BlNV||MK956105:segs',\n",
       " 'TurrNV||KX883629:segs',\n",
       " 'CASV||KJ125489:segs',\n",
       " 'NseV||JQ957874:segs',\n",
       " 'HanaV||JQ957872:segs',\n",
       " 'KADV||KT966495:segs',\n",
       " 'KSaV||KC807171:segs',\n",
       " 'DKNV||AB753015:segs',\n",
       " 'MenoV||JQ957873:segs',\n",
       " 'NDV||DQ458789:segs',\n",
       " 'DiankeV||KY056254:segs',\n",
       " 'OFAV||KT966491:segs',\n",
       " 'YiV-HB-MLV||KY369959:segs',\n",
       " 'InsMV-1||MN714662:segs',\n",
       " 'ACMV||MN961271:segs',\n",
       " 'FOaMV-1||MN714663:segs',\n",
       " 'PSCNV||MH933735:segs',\n",
       " 'NGSAV||MG600024:segs',\n",
       " 'WJHAV||MG600020:segs',\n",
       " 'CharNV||KX883628:segs',\n",
       " 'DecNV||KX883636:segs',\n",
       " 'PagRV||KX883627:segs',\n",
       " 'MrGV||MT907511:segs',\n",
       " 'GAV||AF227196:segs',\n",
       " 'YHV||EU487200:segs',\n",
       " 'YHV-8||KX947267:segs',\n",
       " 'WBV||DQ898157:segs',\n",
       " 'FHMNV||GU002364:segs',\n",
       " 'CSV||KJ681496:segs',\n",
       " 'BoNV||KM589359:segs',\n",
       " 'HpoToV_33749||MG600028:segs',\n",
       " 'SerTV-25||MN161572:segs',\n",
       " 'INTOV||KX883638:segs',\n",
       " 'VCSTV-A||MT997160:segs',\n",
       " 'GRBSToV||MG600030:segs',\n",
       " 'BPNV||KJ541759:segs',\n",
       " 'MVNV||MF351889:segs',\n",
       " 'MoVNV||MK182569:segs',\n",
       " 'BRV||MF685025:segs',\n",
       " 'ShNV||KX184715:segs',\n",
       " 'SECTOV||KX883637:segs',\n",
       " 'SerTV-K48||MN161566:segs',\n",
       " 'SerTV-C18||MN161561:segs',\n",
       " 'VCSTV-B||MT997159:segs',\n",
       " 'BaToV-Hrufipes2018||MW561977:segs',\n",
       " 'BRV||AY427798:segs',\n",
       " 'EToV|0|DQ310701:segs',\n",
       " 'EToV|1|X52506:segs',\n",
       " 'EToV|2|X52505:segs',\n",
       " 'PoTV||JQ860350:segs',\n",
       " 'BoHV||KP641127:segs',\n",
       " 'GHV||KC551800:segs',\n",
       " 'NPHV||KP325401:segs',\n",
       " 'RHV-J||KC411777:segs',\n",
       " 'HCV1a||AF009606:segs',\n",
       " 'HCV1a||M62321:segs',\n",
       " 'HCV1b||D90208:segs',\n",
       " 'HCV2a||D00944:segs',\n",
       " 'HCV2b||D10988:segs',\n",
       " 'HCV3a||D17763:segs',\n",
       " 'HCV3k||D63821:segs',\n",
       " 'HCV4a||GU814265:segs',\n",
       " 'HCV5a||Y13184:segs',\n",
       " 'HCV6a||Y12083:segs',\n",
       " 'HCV6g||D63822:segs',\n",
       " 'HCV7a||EF108306:segs',\n",
       " 'BHV-K||KC796074:segs',\n",
       " 'RHV-F||KC411784:segs',\n",
       " 'NRHV2||KJ950939:segs',\n",
       " 'BHV-M||KC796078:segs',\n",
       " 'RHV-E||KC815310:segs',\n",
       " 'GBV-B||U22304:segs',\n",
       " 'NRHV1||KJ950938:segs',\n",
       " 'NRHV1||KX905133:segs',\n",
       " 'RHV-I||KC411806:segs',\n",
       " 'BHV-L||KC796077:segs',\n",
       " 'APOIV||AF160193:segs',\n",
       " 'AROAV||AY632536:segs',\n",
       " 'AROAV||AF013362:segs',\n",
       " 'BSQV||AF013366:segs',\n",
       " 'IGUV||AF013375:segs',\n",
       " 'NJLV||AF013390:segs',\n",
       " 'BAGV||AY632545:segs',\n",
       " 'BANV||DQ859056:segs',\n",
       " 'BOUV||DQ859057:segs',\n",
       " 'RBV||AF144692:segs',\n",
       " 'BBV||AF013365:segs',\n",
       " 'CPCV||KF917536:segs',\n",
       " 'CIV||AF013368:segs',\n",
       " 'CRV||AF013370:segs',\n",
       " 'DBV||AF013371:segs',\n",
       " 'DENV-2||U87411:segs',\n",
       " 'DENV-1||U88536:segs',\n",
       " 'DENV-3||M93130:segs',\n",
       " 'DENV-4||AF326573:segs',\n",
       " 'EHV||DQ859060:segs',\n",
       " 'TBEV-Eur||U27495:segs',\n",
       " 'TBEV-FE||X07755:segs',\n",
       " 'TBEV-Sib||L40361:segs',\n",
       " 'ENTV||DQ837641:segs',\n",
       " 'SOKV||AF013405:segs',\n",
       " 'YFV||X03700:segs',\n",
       " 'GGYV||DQ235145:segs',\n",
       " 'ILHV||AY632539:segs',\n",
       " 'ROCV||AF013397:segs',\n",
       " 'ITV||AF013377:segs',\n",
       " 'JEV||M18370:segs',\n",
       " 'JUGV||DQ859066:segs',\n",
       " 'JUTV||KJ469371:segs',\n",
       " 'KADV||DQ235146:segs',\n",
       " 'KEDV||AY632540:segs',\n",
       " 'KOKV||AY632541:segs',\n",
       " 'STRV||AF013407:segs',\n",
       " 'KOUV||AF013384:segs',\n",
       " 'KFDV||AY323490:segs',\n",
       " 'AHFV||AF331718:segs',\n",
       " 'LGTV||AF253419:segs',\n",
       " 'SLEV||DQ525916:segs',\n",
       " 'LIV||Y07863:segs',\n",
       " 'LIV-Brit||D12937:segs',\n",
       " 'LIV-Ir||X86784:segs',\n",
       " 'LIV-Spain||DQ235152:segs',\n",
       " 'TSEV||DQ235151:segs',\n",
       " 'GGEV||DQ235153:segs',\n",
       " 'MEAV||DQ235144:segs',\n",
       " 'MODV||AJ242984:segs',\n",
       " 'MMLV||AJ299445:segs',\n",
       " 'MVEV||AF161266:segs',\n",
       " 'ALFV||AF013360:segs',\n",
       " 'WNV||M12294:segs',\n",
       " 'KUNV||D00246:segs',\n",
       " 'NTAV||JX236040:segs',\n",
       " 'OHFV||AY193805:segs',\n",
       " 'SPV||AF013402:segs',\n",
       " 'PPBV||AF013394:segs',\n",
       " 'BCV||AF013369:segs',\n",
       " 'POWV||L06436:segs',\n",
       " 'DTV||AF311056:segs',\n",
       " 'RFV||DQ235149:segs',\n",
       " 'SABV||DQ859062:segs',\n",
       " 'POTV||DQ859067:segs',\n",
       " 'SREV||DQ235150:segs',\n",
       " 'SEPV||DQ837642:segs',\n",
       " 'TMUV||JF895923:segs',\n",
       " 'TYUV||KF815939:segs',\n",
       " 'UGSV||DQ859065:segs',\n",
       " 'USUV||AY453411:segs',\n",
       " 'SVV||AF013401:segs',\n",
       " 'WESSV||EU707555:segs',\n",
       " 'YAOV||AF013413:segs',\n",
       " 'YOKV||AB114858:segs',\n",
       " 'ZIKV||AY632535:segs',\n",
       " 'ZIKV||KY766069:segs',\n",
       " 'ZIKV||KJ776791:segs',\n",
       " 'ZIKV||KX377337:segs',\n",
       " 'EPgV||KC410872:segs',\n",
       " 'BPgV-F||KC796080:segs',\n",
       " 'HHPgV||KT439329:segs',\n",
       " 'TDAV||KC145265:segs',\n",
       " 'HPgV||U44402:segs',\n",
       " 'HPgV||U63715:segs',\n",
       " 'HPgV||AB003289:segs',\n",
       " 'HPgV||U45966:segs',\n",
       " 'HPgV||AY196904:segs',\n",
       " 'HPgV||D87713:segs',\n",
       " 'HPgV||AB021287:segs',\n",
       " 'HPgV||AY949771:segs',\n",
       " 'HPgV||AB003292:segs',\n",
       " 'SPgV-ch||AF070476:segs',\n",
       " 'RPbV||KC815311:segs',\n",
       " 'GBV-A||U22303:segs',\n",
       " 'GBV-A||U94421:segs',\n",
       " 'GBV-D||GU566734:segs',\n",
       " 'GBV-D||GU566735:segs',\n",
       " 'BPgV-G||KC796076:segs',\n",
       " 'BPgV-I||KC796088:segs',\n",
       " 'PPgV||KU351669:segs',\n",
       " 'PAPeV||AY781152:segs',\n",
       " 'PPeV||EF100713:segs',\n",
       " 'AydinPeV||JX428945:segs',\n",
       " 'BVDV1||M96751:segs',\n",
       " 'BVDV1||M31182:segs',\n",
       " 'HoBiPeV||FJ040215:segs',\n",
       " 'GPeV||AF144617:segs',\n",
       " 'LindaV||KY436034:segs',\n",
       " 'PhoPeV||MK910227:segs',\n",
       " 'TSV||MZ664274:segs',\n",
       " 'ovIT PeV||MK618726:segs',\n",
       " 'BDV||AF037405:segs',\n",
       " 'DYPV||MK636874:segs',\n",
       " 'RtNn-PeV||KY370101:segs',\n",
       " 'RtAp-PeV||KY370100:segs',\n",
       " 'RPeV||KJ950914:segs',\n",
       " 'BtSk-PeV||MH282908:segs',\n",
       " 'APPeV||KR011347:segs',\n",
       " 'CSFV||X87939:segs',\n",
       " 'BVDV2||FJ527854:segs',\n",
       " 'BVDV2||U18059:segs',\n",
       " 'LLOV||JF828358:segs',\n",
       " 'MLAV||KX371887:segs',\n",
       " 'OBLV||MN510773:segs',\n",
       " 'BOMV||MF319185:segs',\n",
       " 'BDBV||FJ217161:segs',\n",
       " 'RESTV||AF522874:segs',\n",
       " 'SUDV||AY729654:segs',\n",
       " 'TAFV||FJ217162:segs',\n",
       " 'EBOV||AF086833:segs',\n",
       " 'MARV||DQ217792:segs',\n",
       " 'RAVV||DQ447649:segs',\n",
       " 'XILV||MG599980:segs',\n",
       " 'TAPV||BR001752:segs',\n",
       " 'KNDV||MW093492:segs',\n",
       " 'FIWIV||MN510772:segs',\n",
       " 'HUJV||MG599981:segs',\n",
       " 'APMV8||FJ215863:segs',\n",
       " 'APMV10||HM147142:segs',\n",
       " 'APMV11||JQ886184:segs',\n",
       " 'APMV6||EF569970:segs',\n",
       " 'APMV14||KX258200:segs',\n",
       " 'APMV20||MF033136:segs',\n",
       " 'APMV5||GU206351:segs',\n",
       " 'APMV15||KX932454:segs',\n",
       " 'APMV22||MK677430:segs',\n",
       " 'APMV7||FJ231524:segs',\n",
       " 'APMV2||EU338414:segs',\n",
       " 'APVA||KY452442:segs',\n",
       " 'APMV12||KC333050:segs',\n",
       " 'APMV13||LC041132:segs',\n",
       " 'APMV1||AY562991:segs',\n",
       " 'APVB||KY452443:segs',\n",
       " 'APMV21||MF594598:segs',\n",
       " 'APMV9||EU910942:segs',\n",
       " 'APVC||KY452444:segs',\n",
       " 'APMV16||KY511044:segs',\n",
       " 'APMV4||EU877976:segs',\n",
       " 'APMV3||EU782025:segs',\n",
       " 'WTLPV||MG600058:segs',\n",
       " 'PSPV||MH900516:segs',\n",
       " 'ASPV||EU156171:segs',\n",
       " 'FDLV||AY141760:segs',\n",
       " 'CedV||JQ001776:segs',\n",
       " 'GhV||HQ660129:segs',\n",
       " 'HeV||AF017149:segs',\n",
       " 'MojV||KF278639:segs',\n",
       " 'NiV||AF212302:segs',\n",
       " 'MisPV||KC154054:segs',\n",
       " 'RoPV||KY370098:segs',\n",
       " 'BeiV||DQ100461:segs',\n",
       " 'BatPV1||MG203877:segs',\n",
       " 'BeV||MN561699:segs',\n",
       " 'FPaV||LC431581:segs',\n",
       " 'MMLV1||MG573140:segs',\n",
       " 'MMLV2||MG573141:segs',\n",
       " 'BatPV2||MG203878:segs',\n",
       " 'ShaV||MG230624:segs',\n",
       " 'BatPV3||KJ641657:segs',\n",
       " 'PMPV1||MG516455:segs',\n",
       " 'JV||AY900001:segs',\n",
       " 'RulV||MW579602:segs',\n",
       " 'TaiV||JN689227:segs',\n",
       " 'CDV||AF014953:segs',\n",
       " 'PPRV||AJ849636:segs',\n",
       " 'CeMV||AJ608288:segs',\n",
       " 'FeMV||JQ411014:segs',\n",
       " 'MeV||AB016162:segs',\n",
       " 'RPV||X98291:segs',\n",
       " 'PDV||KC802221:segs',\n",
       " 'MossV||AY286409:segs',\n",
       " 'BaV1||MF943130:segs',\n",
       " 'NarV||FJ362497:segs',\n",
       " 'TupV||AF079780:segs',\n",
       " 'BPIV3||AF178654:segs',\n",
       " 'CPIV3||KT215610:segs',\n",
       " 'HPIV1||AF457102:segs',\n",
       " 'SeV||AB195968:segs',\n",
       " 'HPIV3||EU326526:segs',\n",
       " 'GSqV||LS992584:segs',\n",
       " 'PPIV1||JX857409:segs',\n",
       " 'SalV||JQ697837:segs',\n",
       " 'AlsV||MH972568:segs',\n",
       " 'HPIV4a||AB543336:segs',\n",
       " 'HPIV4b||AB543337:segs',\n",
       " 'HPIV2||X57559:segs',\n",
       " 'PIV5||AF052755:segs',\n",
       " 'MapV||EF095490:segs',\n",
       " 'MuV||AB040874:segs',\n",
       " 'SV41||X64275:segs',\n",
       " 'LPMV||BK005918:segs',\n",
       " 'AchPV2||JX051320:segs',\n",
       " 'AchPV1||JX051319:segs',\n",
       " 'ThkPV2||GU128081:segs',\n",
       " 'ThkPV1||GU128080:segs',\n",
       " 'HerV||KU672593:segs',\n",
       " 'ThkPV3||GU128082:segs',\n",
       " 'MenPV||JX112711:segs',\n",
       " 'SOSV||KF774436:segs',\n",
       " 'TevPV||KP271124:segs',\n",
       " 'TioPV||AF298895:segs',\n",
       " 'WTSPV||MG600059:segs',\n",
       " 'WHPV||MG600062:segs',\n",
       " 'WPSSPV||MG600057:segs']"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[i[0].stem for i in search_result]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## note for OSV-1||MF324848-origin\n",
    "### case1. the clan nearby should include this two entry.\n",
    "1. Orbi_VP1:\n",
    "    https://www.ebi.ac.uk/interpro/entry/pfam/PF05788/\n",
    "    Also RdRp\n",
    "\n",
    "2. IFR3_antag:\n",
    "    https://www.ebi.ac.uk/interpro/entry/pfam/PF14754/\n",
    "    Also proteinase, and hit prob=100% here.\n",
    "\n",
    "### case2 accessory domains of nearby function domain\n",
    "1. Hybird (Rep_1B)\n",
    "    https://www.ebi.ac.uk/interpro/entry/pfam/PF17873/\n",
    "    \"a regulatory domain found in replicase polyprotein 1ab found in Arterivirus\"\n",
    "\n",
    "    hybrid:\n",
    "    This superfamily contains proteins with a hybrid motif embedded in structurally diverse proteins.\n",
    "    PMID: 8081735\n",
    "\n",
    "2. DUF3388\n",
    "    https://www.ebi.ac.uk/interpro/entry/pfam/PF11868/\n",
    "    'functionally uncharacterised. This protein is found in bacteria'\n",
    "    'associated with Pfam:PF01842: ACT domain'\n",
    "\n",
    "3. RING (zf-RING_13)\n",
    "    https://www.ebi.ac.uk/interpro/entry/pfam/PF17977/\n",
    "    RING/Ubox like zinc-binding domain\n",
    "\n",
    "### case3 valid multiple hits\n",
    "1. Arteri_Gl\n",
    "    https://www.ebi.ac.uk/interpro/entry/pfam/PF00951/\n",
    "    Gl envelope protein encoded by Arteriviruses \n",
    "\n",
    "### case4 multiple hits should be integrated\n",
    "1. Peptidase_S46\n",
    "    https://www.ebi.ac.uk/interpro/entry/pfam/PF10459/\n",
    "    member of Peptidase_PA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "domain_annotation\n",
       "Arteri_Gl        2\n",
       "Peptidase_S46    2\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hits_df['domain_annotation'].value_counts()[hits_df['domain_annotation'].value_counts()>1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>89</th>\n",
       "      <th>94</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>domain_accession</th>\n",
       "      <td>PF10459</td>\n",
       "      <td>PF10459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>domain_annotation</th>\n",
       "      <td>Peptidase_S46</td>\n",
       "      <td>Peptidase_S46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prob</th>\n",
       "      <td>95.2</td>\n",
       "      <td>74.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eval</th>\n",
       "      <td>0.00011</td>\n",
       "      <td>0.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pval</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>score</th>\n",
       "      <td>84.0</td>\n",
       "      <td>57.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ss</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>begin</th>\n",
       "      <td>1393</td>\n",
       "      <td>1487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>end</th>\n",
       "      <td>1418</td>\n",
       "      <td>1513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hmmbegin</th>\n",
       "      <td>43</td>\n",
       "      <td>640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hmmend</th>\n",
       "      <td>66</td>\n",
       "      <td>666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hmmlength</th>\n",
       "      <td>709</td>\n",
       "      <td>709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transannot</th>\n",
       "      <td>1s</td>\n",
       "      <td>1s</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seg_b</th>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seg_e</th>\n",
       "      <td>2091</td>\n",
       "      <td>2091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gen_b</th>\n",
       "      <td>4249</td>\n",
       "      <td>4531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gen_e</th>\n",
       "      <td>4324</td>\n",
       "      <td>4609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gen_hmm_b</th>\n",
       "      <td>4123</td>\n",
       "      <td>2614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gen_hmm_e</th>\n",
       "      <td>6253</td>\n",
       "      <td>4738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clan_accession</th>\n",
       "      <td>CL0124</td>\n",
       "      <td>CL0124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clan_annotation</th>\n",
       "      <td>Peptidase_PA</td>\n",
       "      <td>Peptidase_PA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              89             94\n",
       "domain_accession         PF10459        PF10459\n",
       "domain_annotation  Peptidase_S46  Peptidase_S46\n",
       "prob                        95.2           74.0\n",
       "eval                     0.00011           0.12\n",
       "pval                         0.0       0.000028\n",
       "score                       84.0           57.3\n",
       "ss                           0.0            0.0\n",
       "begin                       1393           1487\n",
       "end                         1418           1513\n",
       "hmmbegin                      43            640\n",
       "hmmend                        66            666\n",
       "hmmlength                    709            709\n",
       "transannot                    1s             1s\n",
       "seg_b                         23             23\n",
       "seg_e                       2091           2091\n",
       "gen_b                       4249           4531\n",
       "gen_e                       4324           4609\n",
       "gen_hmm_b                   4123           2614\n",
       "gen_hmm_e                   6253           4738\n",
       "clan_accession            CL0124         CL0124\n",
       "clan_annotation     Peptidase_PA   Peptidase_PA"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hits_df[hits_df['domain_annotation']=='Peptidase_S46'].T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>96</th>\n",
       "      <th>99</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>domain_accession</th>\n",
       "      <td>PF00951</td>\n",
       "      <td>PF00951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>domain_annotation</th>\n",
       "      <td>Arteri_Gl</td>\n",
       "      <td>Arteri_Gl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prob</th>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eval</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pval</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>score</th>\n",
       "      <td>344.7</td>\n",
       "      <td>432.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ss</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>begin</th>\n",
       "      <td>17</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>end</th>\n",
       "      <td>182</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hmmbegin</th>\n",
       "      <td>18</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hmmend</th>\n",
       "      <td>172</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hmmlength</th>\n",
       "      <td>173</td>\n",
       "      <td>173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transannot</th>\n",
       "      <td>1s</td>\n",
       "      <td>2s</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seg_b</th>\n",
       "      <td>4247</td>\n",
       "      <td>4012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seg_e</th>\n",
       "      <td>4430</td>\n",
       "      <td>4257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gen_b</th>\n",
       "      <td>12793</td>\n",
       "      <td>12269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gen_e</th>\n",
       "      <td>13288</td>\n",
       "      <td>12770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gen_hmm_b</th>\n",
       "      <td>12742</td>\n",
       "      <td>12254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gen_hmm_e</th>\n",
       "      <td>13291</td>\n",
       "      <td>12773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clan_accession</th>\n",
       "      <td>PF00951</td>\n",
       "      <td>PF00951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clan_annotation</th>\n",
       "      <td>Arteri_Gl</td>\n",
       "      <td>Arteri_Gl</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          96         99\n",
       "domain_accession     PF00951    PF00951\n",
       "domain_annotation  Arteri_Gl  Arteri_Gl\n",
       "prob                   100.0      100.0\n",
       "eval                     0.0        0.0\n",
       "pval                     0.0        0.0\n",
       "score                  344.7      432.9\n",
       "ss                       0.0        0.0\n",
       "begin                     17         77\n",
       "end                      182        244\n",
       "hmmbegin                  18          6\n",
       "hmmend                   172        172\n",
       "hmmlength                173        173\n",
       "transannot                1s         2s\n",
       "seg_b                   4247       4012\n",
       "seg_e                   4430       4257\n",
       "gen_b                  12793      12269\n",
       "gen_e                  13288      12770\n",
       "gen_hmm_b              12742      12254\n",
       "gen_hmm_e              13291      12773\n",
       "clan_accession       PF00951    PF00951\n",
       "clan_annotation    Arteri_Gl  Arteri_Gl"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hits_df[hits_df['clan_annotation']=='Arteri_Gl'].T"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
